src/client.rs

```diff
@@ -152,48 +152,44 @@ where
         let mut stream =
             Stream::new(&mut this.io, &mut this.session).set_eof(!this.state.readable());
 
-        #[allow(clippy::match_single_binding)]
-        match this.state {
-            #[cfg(feature = "early-data")]
-            TlsState::EarlyData(ref mut pos, ref mut data) => {
-                use std::io::Write;
-
-                // write early data
-                if let Some(mut early_data) = stream.session.early_data() {
-                    let len = match early_data.write(buf) {
-                        Ok(n) => n,
-                        Err(err) => return Poll::Ready(Err(err)),
-                    };
-                    if len != 0 {
-                        data.extend_from_slice(&buf[..len]);
-                        return Poll::Ready(Ok(len));
-                    }
+        #[cfg(feature = "early-data")]
+        if let TlsState::EarlyData(pos, data) = &mut this.state {
```

Comment:
That's better. I was hoping for something that reduces indentation for the early data state one more level, but I suppose that's tricky to square with the conditional compilation guard...

@quininer do you still think it makes sense to have a Cargo feature for early data? What would the cost be of always enabling it?

Reply:
early data has extra overhead and different behavior, especially 0-RTT which actually reduces security. I think it is good to have feature gate.